languages/javascript/exercises/concept/errors/.docs/instructions.md

```diff
@@ -0,0 +1,26 @@
+You are a developer in a newspaper factory and you are asked to work on a software that prevents  printing machines from overheating. Currently there are technicians that check each hour the temperatures and turn of the machines if they are overheating.
+
+## Detect overheating
+
+You are provided with a Javascript API for a machine that returns its current status information.
+
+You have been told by the technicians that the machines' temperature should never exceed 500 °C, because they could catch fire if they did.
+
+When the temperature is higher than that, you should throw an error with a message containing the current temperature.
+
+## Shutdown
+
+You now need to catch that error and shutdown the machine.
```

Comment:
It's going to be awkward to call `shutdown` on an API, when it also provides the temperature. Here is my iteration based on your story. You **don't need to use this**, but you may! Just ideas.

```markdown
## Detect overheating

[...]

Your task is to implement a function `reportOverheating` that receives a 
temperature from one of the sensors. If the temperature exceeds 500° C., 
throw an error. 
```
Doesn't introduce custom errors yet.

```markdown
## Input validation

A few of the machines are an older model, and unfortunately they don't always
report the temperature. When there is no temperature reported, throw an error. 
```
Teaches that errors can also be used for input validation / fail-fast. We would have an _analyzer_ test if there is a `return` statement after the throwing, and give a message "unreachable `return` detected".

```markdown
## Making the error actionable

In order to correctly respond to issues, the function should report different 
errors for different cases. Change the function to throw an 
`ArgumentError` error when the input is invalid, and an `OverheatingError` 
when the temperature exceeds the limit. The `OverheatingError` should have 
a property `.temperature` which reports the temperature that was given.
```
Introduces custom errors (and that they can have properties). This should be added to the dependencies (`classes` and `properties` or something like that).

```markdown
## Catching issues

Your boss is very happy with your work but has noticed an issue. The engineer
responsible for implementing the system that can prevent further damage has
not done their job right. 

Implement a function `guardTheNews` that receives an argument `check`, and 
an argument `shutdown`. 

- `check` is a _function_ that, when called, checks the sensors of the machine. 
  It may throw various errors.
- `shutdown` is a function that protects the machine from blowing up.

The function `guardTheNews` should call `check()`. If it passes, the function
should return `"Everything is fine."`. However, it may `throw` one of the 
following errors:

- `ArgumentError`: when this happens, retry
- `OverheatingError`: when this happens, shut down
- _anything else_: when this happens, rethrow the error
```
Introduces `try`, `catch`. Add `callbacks` to the dependencies (even tho the nomenclature is meh here -- function as a first class citizen is the right term). `instanceof` should have been part of `classes`, so that's fine.

Reply:
Thanks for the ideas, I like the idea of old machines that you'd have to ask the temperatures to multiple times  👍 

I'll gather the ideas in a new draft

Reply:
@SleeplessByte I've updated the instructions' file.

I am wondering about the exercise steps. The fact that first you throw errors then have to come back on your code and throw Error classes. How would that work in real life ? Should we provide an exercise file with something like `functionVersion1` and `functionVersion2`.